app/scripts/metamask-controller.js

```diff
@@ -185,10 +184,25 @@ module.exports = class MetamaskController {
       },
       // tx signing
       approveTransaction: this.newUnsignedTransaction.bind(this),
-      signTransaction: (...args) => {
-        this.setupSigningListners(...args)
-        this.txManager.formatTxForSigining(...args)
-        this.sendUpdate()
+      signTransaction: (txParams, cb) => {
+        this.txManager.formatTxForSigining(txParams)
+        .then(({ethTx, address, txId}) => {
+          return this.keyringController.signTransaction(ethTx, address, txId)
+        })
+        .then(({tx, txId}) => {
+          return this.txManager.resolveSignedTransaction({tx, txId})
+        })
+        .then((rawTx) => {
+          cb(null, rawTx)
+        })
+        .catch((err) => {
+          console.error(err)
+          cb(err)
+        })
+        .then(() => {
+          this.sendUpdate()
+          this.txManager.emit(`${txParams.metamaskId}:signingComplete`)
+        })
```

Comment:
This is my one comment that I'd like actually changed.

There seem to be two final "then" blocks.  One on `195` and one on `202`.

You should probably be able to combine them into one block.  The `catch` between them shouldn't influence their flow at all, since it's only triggered in the case of an error.